rts/motoko-rts/src/idl.rs

```diff
@@ -486,3 +512,335 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
+                    let t21 = sleb128_decode(&mut tb2);
+                    if n1 == 0 {
+                        // check all remaining fields optional
+                        if !is_null_opt_reserved(typtbl2, end2, t21) {
+                            break 'return_false;
+                        }
+                        continue;
+                    };
+                    if advance {
+                        loop {
+                            tag1 = leb128_decode(&mut tb1);
+                            t11 = sleb128_decode(&mut tb1);
+                            n1 -= 1;
+                            if !(tag1 < tag2 && n1 > 0) {
+                                break;
+                            }
+                        }
+                    };
+                    if tag1 > tag2 {
+                        if !is_null_opt_reserved(typtbl2, end2, t21) {
+                            // missing, non_opt field
+                            break 'return_false;
+                        }
+                        advance = false; // reconsider this field in next round
+                        continue;
+                    };
+                    if !sub(rel, p, typtbl1, typtbl2, end1, end2, t11, t21) {
+                        break 'return_false;
+                    }
+                    advance = true;
+                }
+                return true;
+            }
+            (IDL_CON_variant, IDL_CON_variant) => {
+                let n1 = leb128_decode(&mut tb1);
+                let mut n2 = leb128_decode(&mut tb2);
+                for _ in 0..n1 {
+                    if n2 == 0 {
+                        break 'return_false;
+                    };
+                    let tag1 = leb128_decode(&mut tb1);
+                    let t11 = sleb128_decode(&mut tb1);
+                    let mut tag2: u32;
+                    let mut t21: i32;
+                    loop {
+                        tag2 = leb128_decode(&mut tb2);
+                        t21 = sleb128_decode(&mut tb2);
+                        n2 -= 1;
+                        if !(tag2 < tag1 && n2 > 0) {
+                            break;
+                        }
+                    }
+                    if tag1 != tag2 {
+                        break 'return_false;
+                    };
+                    if !sub(rel, p, typtbl1, typtbl2, end1, end2, t11, t21) {
+                        break 'return_false;
+                    }
+                }
+                return true;
+            }
+            (IDL_CON_service, IDL_CON_service) => {
+                let mut n1 = leb128_decode(&mut tb1);
+                let n2 = leb128_decode(&mut tb2);
+                for _ in 0..n2 {
+                    if n1 == 0 {
+                        break 'return_false;
+                    };
+                    let len2 = leb128_decode(&mut tb2);
+                    let p2 = tb2.ptr;
+                    Buf::advance(&mut tb2, len2);
```

Comment:
Worth abstracting over this, with a function taking `tb` and returning `(len, ptr)`?

Reply:
Done